fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LucenOnlineIndexingTest.java

```diff
@@ -0,0 +1,631 @@
+            }
+        }
+        return needMerge;
+    }
+
+    @SuppressWarnings("checkstyle:VariableDeclarationUsageDistance")
+    @Test
+    void testRecordUpdateBackgroundMerge2() {
+        Index index = SIMPLE_TEXT_SUFFIXES;
+        boolean needMerge = populateDataSplitSegments(index, 20, 5);
+        assertTrue(needMerge);
+
+        String[] allFiles = listFiles(index);
+        int oldLength = allFiles.length;
+
+        recordStore.getIndexDeferredMaintenancePolicy().setAutoMergeDuringCommit(false);
+        try (OnlineIndexer indexBuilder = OnlineIndexer.newBuilder()
+                .setRecordStore(recordStore)
+                .setIndex(index)
+                .build()) {
+            indexBuilder.mergeIndex();
+        }
+
+        allFiles = listFiles(index);
+        int newLength = allFiles.length;
+        LOGGER.debug("Merge test: number of files: old=" + oldLength + " new=" + newLength);
+        assertTrue(newLength < oldLength);
+        oldLength = newLength;
+
+        recordStore.getIndexDeferredMaintenancePolicy().setAutoMergeDuringCommit(false);
+        try (OnlineIndexer indexBuilder = OnlineIndexer.newBuilder()
+                .setRecordStore(recordStore)
+                .setIndex(index)
+                .build()) {
+            indexBuilder.mergeIndex();
+        }
+
+        allFiles = listFiles(index);
+        newLength = allFiles.length;
+        LOGGER.debug("Merge test: number of files: old=" + oldLength + " new=" + newLength);
+        assertTrue(newLength < oldLength);
+    }
+
+    @SuppressWarnings("checkstyle:VariableDeclarationUsageDistance")
+    @Test
+    void testRecordUpdateMergeByAnotherRecordUpdate() {
+        Index index = SIMPLE_TEXT_SUFFIXES;
+        // update records while skipping merge
+        boolean needMerge = populateDataSplitSegments(index, 20, 5);
+        assertTrue(needMerge);
+
+        String[] allFiles = listFiles(index);
+        int oldLength = allFiles.length;
+
+        // Now update records with merge
+        try (FDBRecordContext context = openContext()) {
+            rebuildIndexMetaData(context, SIMPLE_DOC, index);
+            recordStore.getIndexDeferredMaintenancePolicy().setAutoMergeDuringCommit(true);
+            for (int i = 17; i < 22; i++) {
+                recordStore.saveRecord(createSimpleDocument(1623L + i, ENGINEER_JOKE + " Sababa", 2));
+            }
+            commit(context);
+        }
+
+        allFiles = listFiles(index);
+        int newLength = allFiles.length;
+        LOGGER.debug("Merge test: number of files: old=" + oldLength + " new=" + newLength);
+        assertTrue(newLength < oldLength);
+    }
+
+    @Test
+    void testRecordUpdateReducedMerge() {
+        // emulate repeating merge until until unchanged.
+        Index index = SIMPLE_TEXT_SUFFIXES;
+        boolean needMerge = populateDataSplitSegments(index, 40, 7);
+        assertTrue(needMerge);
+
+        boolean allDone = false;
+        while (! allDone) {
```

Comment:
It seems worthwhile to assert that it goes through this loop at least once, maybe at least twice.

Reply:
✅